tests/data_context/store/test_datasource_store.py

```diff
@@ -0,0 +1,158 @@
+from unittest.mock import PropertyMock, patch
+
+import pytest
+
+from great_expectations.core.data_context_key import StringKey
+from great_expectations.data_context.store.datasource_store import DatasourceStore
+from great_expectations.data_context.types.base import (
+    DatasourceConfig,
+    DatasourceConfigSchema,
+)
+from great_expectations.data_context.types.resource_identifiers import GeCloudIdentifier
+
+
+@pytest.fixture
+def datasource_name() -> str:
+    return "my_first_datasource"
+
+
+@pytest.fixture
+def datasource_store_name() -> str:
+    return "datasource_store"
+
+
+@pytest.fixture
+def datasource_key(datasource_name: str) -> StringKey:
+    return StringKey(key=datasource_name)
+
+
+@pytest.fixture
+def empty_datasource_store(datasource_store_name: str) -> DatasourceStore:
+    return DatasourceStore(store_name=datasource_store_name)
+
+
+@pytest.fixture
+def datasource_config() -> DatasourceConfig:
+    return DatasourceConfig(
+        class_name="Datasource",
+        execution_engine={
+            "class_name": "PandasExecutionEngine",
+            "module_name": "great_expectations.execution_engine",
+        },
+        data_connectors={
+            "tripdata_monthly_configured": {
+                "class_name": "ConfiguredAssetFilesystemDataConnector",
+                "module_name": "great_expectations.datasource.data_connector",
+                "base_directory": "/path/to/trip_data",
+                "assets": {
+                    "yellow": {
+                        "class_name": "Asset",
+                        "module_name": "great_expectations.datasource.data_connector.asset",
+                        "pattern": r"yellow_tripdata_(\d{4})-(\d{2})\.csv$",
+                        "group_names": ["year", "month"],
+                    }
+                },
+            }
+        },
+    )
+
+
+def test_datasource_store_with_bad_key_raises_error(
+    empty_datasource_store: DatasourceStore, datasource_config: DatasourceConfig
+) -> None:
+    store: DatasourceStore = empty_datasource_store
+
+    error_msg: str = "key must be an instance of StringKey"
+
+    with pytest.raises(TypeError) as e:
+        store.set(key="my_bad_key", value=datasource_config)
+    assert error_msg in str(e.value)
+
+    with pytest.raises(TypeError) as e:
+        store.get(key="my_bad_key")
+    assert error_msg in str(e.value)
+
+
+def test_datasource_store_retrieval(
+    empty_datasource_store: DatasourceStore, datasource_config: DatasourceConfig
+) -> None:
+    store: DatasourceStore = empty_datasource_store
+
+    key: StringKey = StringKey(key="my_datasource")
+    store.set(key=key, value=datasource_config)
+    res: DatasourceConfig = store.get(key=key)
+
+    assert isinstance(res, DatasourceConfig)
+    assert DatasourceConfigSchema().dump(res) == datasource_config.to_dict()
```

Comment:
asserting the key here should be sufficient?

Reply:
I also don't think that `DatasourceConfig` has implemented `__eq__` so it will compare by memory address instead of actual values.

Reply:
sorry for not being clear, I meant: could we assert key values for equality vs. the whole payload and still be testing what we intend to be